### ISSUE TITLE
Crash when inserting object having related object that already exists

### ISSUE THREAD

**Issue by yasirmturk** [realm/realm-cocoa#2590]
i have an Entity A, it does not have a primary key and it has a property 
`@property RLMArray<City> *Cities;` but City has a primary key `NSString *Id;` 
I am inserting array of As using `[A createInRealm:realm withValue:aDict];`, where `aDict[@"Cities"]` is an array of Cities which is most likely to contain duplicate objects across different objects of A, as most of the graph serializers generate the JSON. inside the code it is always trying to insert-only the nested object City instead of upsert, hence generating this expection. I think it was working fine with v 0.91.5

`Terminating app due to uncaught exception 'RLMException', reason: 'Can't set primary key property 'Id' to existing value '6200'.'`
in 
`RLMSetValueUnique
...
RLMGetLinkedObjectForValue 
...
RLMCreateObjectInRealmWithValue`

**Comment by jpsim** [comment]
At the moment, you can't mix and match models with and without primary keys when using `createOrUpdate`. It's fundamentally opposed to the concept of identity that primary keys provide.

If you need a custom identity management approach, you could break up your creation logic into more steps (e.g. `createInRealm:withValue:` followed by setting the `cities` property with existing values).

**Comment by yasirmturk** [comment]
that could be a work-around but not the proper way to address this problem. We are talking an ORM and Not having a primary key is not fundamentally wrong. 

As i see it, while adding the Linked attributes it should ALWAYS used upsert IF the linked object class has a primary key.

**Comment by jpsim** [comment]
We've discussed changing `createInRealm:withValue:` to create or update nested objects when they have primary keys before (although I can't find the relevant GH issue at the moment).

We opted against that in order to maintain the guarantee that calling create (without update) will always create objects, not update them.

However, we could maintain this guarantee if we extended `createOrUpdate:` to allow being called for root objects without a primary key, which would update its linked objects if they existed and already have the same primary key. Thoughts on this @realm/cocoa?

PS: Minor point, but Realm is not an ORM. There's no relational mapping involved, as underlying db engine stores relationships natively.

**Comment by yasirmturk** [comment]
i have just come across a line 
`return RLMCreateObjectInRealmWithValue(realm, className, value, creationOptions & RLMCreationOptionsCreateOrUpdate);`
in method `RLMGetLinkedObjectForValue`, where i guess the AND op in last parameter `creationOptions & RLMCreationOptionsCreateOrUpdate` has NO effect, however if ORed, it might make sense.. any idea what is it meant for? 

P.S. if its not an ORM how can i define all kind of relationships without specifying any primary or foreign keys and any intermediate tables

**Comment by jpsim** [comment]
> P.S. if its not an ORM how can i define all kind of relationships without specifying any primary or foreign keys and any intermediate tables

Aha! You've uncovered the magic of Realm :tophat: :tada:. But seriously, the database file format represents object links (relationships) natively, without the need for expensive primary/foreign key mappings or "join"-style operations constructing intermediate tables.

Back to your real question: that bitwise AND converts `createOptions` to a boolean value for whether or not to "create or update", so it *does* in fact have an effect.

**Comment by yasirmturk** [comment]
So, we can not change AND to OR?

**Comment by jpsim** [comment]
> So, we can not change AND to OR?

Not without sacrificing the guarantee that calling create (without update) will always create objects, not update them.

**Comment by yasirmturk** [comment]
Strange guarantee it is... It actually guarantees the crash for all to-many relationships if linked object have a primary key.. because objects will definitely repeat in to-many relationships..

**Comment by jpsim** [comment]
> Strange guarantee it is... It actually guarantees the crash for all to-many relationships if linked object have a primary key.. because objects will definitely repeat in to-many relationships..

Before this conversation continues on this path any further, would the solution I outlined in one of my earlier comments be acceptable to you?

> However, we could maintain this guarantee if we extended `createOrUpdate:` to allow being called for root objects without a primary key, which would update its linked objects if they existed and already have the same primary key.

**Comment by yasirmturk** [comment]
I have already implemented the workaround to remove the relation from JSON and manually process it.. but i dont think that is a proper solution.
`[converted removeObjectForKey:@"Cities"];
rlmObject = [A createInRealm:realm  withValue:converted];
for (NSDictionary *city in original[@"Cities"]) {
                [rlmObject.Cities addObject:[City createOrUpdateInRealm:realm withValue:city]];
}
`

**Comment by SandyChapman** [comment]
@jpsim : Stepping into this conversation since I've come looking for the same issue. My situation:

I'm using code generation to generate a Swift client library from a RESTful Swagger document. This is producing a bunch of code where after a REST call is made, the response is parsed into a bunch of nested objects which conform to Realms requirements and inherit from RealmSwift.Object.

At the top level of most of my calls is a hypermedia based object which supports pagination. This does not have a primary key, but the content embedded within the paginated object does have primary keys. I'd prefer to call "add(:update:)" on the paginated object which ignores the fact that some objects may not have primary keys and thus will have no effect rather that iterate through a potentially complex data structure adding / update keys manually.

So in short, yes, I think your suggestion that the behaviour described for `createOrUpdate:` would make sense.

**Comment by jpsim** [comment]
Alright, I'm glad we could reach an agreement on this! I've created #2632 to track this feature. Thanks for your input @yasirmturk and @SandyChapman.
